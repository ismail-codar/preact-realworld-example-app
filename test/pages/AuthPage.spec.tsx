import { h } from 'preact';
import { fireEvent, render, screen } from '@testing-library/preact';

import AuthPage from '../../src/pages/AuthPage';

describe('AuthPage renders', () => {
	it('the Login page correctly', () => {
		const { asFragment } = render(<AuthPage />);
		expect(
			screen.getByRole('heading', { name: 'Sign in' })
		).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it('the Register page correctly', () => {
		const { asFragment } = render(<AuthPage isRegister />);
		expect(
			screen.getByRole('heading', { name: 'Sign up' })
		).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it('the Login input fields', () => {
		render(<AuthPage />);
		expect(screen.getByRole('textbox', { name: 'Email' })).toBeInTheDocument();
		// input[type="password"] doesn't actually have an implicit role,
		// so the test has to be a bit different. See:
		// https://github.com/testing-library/dom-testing-library/issues/567
		expect(screen.getByPlaceholderText('Password')).toBeInTheDocument();
	});

	it('the Register input fields', () => {
		render(<AuthPage isRegister />);
		expect(
			screen.getByRole('textbox', { name: 'Username' })
		).toBeInTheDocument();
		expect(screen.getByRole('textbox', { name: 'Email' })).toBeInTheDocument();
		// input[type="password"] doesn't actually have an implicit role,
		// so the test has to be a bit different. See:
		// https://github.com/testing-library/dom-testing-library/issues/567
		expect(screen.getByPlaceholderText('Password')).toBeInTheDocument();
	});

	it('a (disabled) Login button', () => {
		render(<AuthPage />);
		const submitButton = screen.getByRole('button', { name: 'Sign in' });
		expect(submitButton).toBeInTheDocument();
		expect(submitButton).toBeDisabled();
	});

	it('a (disabled) Register button', () => {
		render(<AuthPage isRegister />);
		const submitButton = screen.getByRole('button', { name: 'Sign up' });
		expect(submitButton).toBeInTheDocument();
		expect(submitButton).toBeDisabled();
	});
});

describe('AuthPage form validates', () => {
	it('username input', () => {
		render(<AuthPage isRegister />);
		const username = screen.getByRole('textbox', { name: 'Username' });

		fireEvent.input(username, { target: { value: '' } });
		expect(username).toBeInvalid();

		fireEvent.input(username, { target: { value: 'Foo' } });
		expect(username).toBeValid();
	});

	it('email input', () => {
		render(<AuthPage />);
		const email = screen.getByRole('textbox', { name: 'Email' });

		fireEvent.input(email, { target: { value: 'smoketest' } });
		expect(email).toBeInvalid();

		fireEvent.input(email, { target: { value: 'smoketest@example.com' } });
		expect(email).toBeValid();
	});

	it('password input', () => {
		render(<AuthPage />);
		const password = screen.getByPlaceholderText('Password');

		fireEvent.input(password, { target: { value: 'foobar' } });
		expect(password).toBeInvalid();

		fireEvent.input(password, { target: { value: 'foobarbaz' } });
		expect(password).toBeValid();
	});

	it('login submission requirements', () => {
		render(<AuthPage />);
		const submitButton = screen.getByRole('button', { name: 'Sign in' });
		expect(submitButton).toBeDisabled();

		fireEvent.input(screen.getByRole('textbox', { name: 'Email' }), {
			target: { value: 'smoketest@example.com' }
		});
		expect(submitButton).toBeDisabled();

		fireEvent.input(screen.getByPlaceholderText('Password'), {
			target: { value: 'foobarbaz' }
		});
		expect(submitButton).toBeEnabled();
	});

	it('registration submission requirements', () => {
		render(<AuthPage isRegister />);
		const submitButton = screen.getByRole('button', { name: 'Sign up' });
		expect(submitButton).toBeDisabled();

		fireEvent.input(screen.getByRole('textbox', { name: 'Username' }), {
			target: { value: 'SmokeTest' }
		});
		expect(submitButton).toBeDisabled();

		fireEvent.input(screen.getByRole('textbox', { name: 'Email' }), {
			target: { value: 'smoketest@example.com' }
		});
		expect(submitButton).toBeDisabled();

		fireEvent.input(screen.getByPlaceholderText('Password'), {
			target: { value: 'foobarbaz' }
		});
		expect(submitButton).toBeEnabled();
	});
});
