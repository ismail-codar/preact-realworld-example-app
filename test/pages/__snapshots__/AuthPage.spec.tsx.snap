// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthPage renders the Login page correctly 1`] = `
<DocumentFragment>
  <div
    class="auth-page"
  >
    <div
      class="container page"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6 offset-md-3 col-xs-12"
        >
          <h1
            class="text-xs-center"
          >
            Sign in
          </h1>
          <p
            class="text-xs-center"
          >
            <a
              href="/register"
            >
              Need an account?
            </a>
          </p>
          <form>
            <fieldset
              class="form-group"
            >
              <input
                aria-label="Email"
                class="form-control form-control-lg"
                placeholder="Email"
                required=""
                type="email"
              />
            </fieldset>
            <fieldset
              class="form-group"
            >
              <input
                aria-label="Password"
                class="form-control form-control-lg"
                minlength="8"
                pattern=".{8,}"
                placeholder="Password"
                required=""
                type="password"
              />
            </fieldset>
            <button
              class="btn btn-lg btn-primary pull-xs-right"
              disabled=""
            >
              Sign in
              <svg
                id="loading-spinner"
                style="margin-left: 0.5rem; display: none;"
                viewBox="0 0 100 100"
                width="1rem"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="50"
                  cy="50"
                  id="circle"
                  r="45"
                  style="stroke: #fff;"
                />
              </svg>
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AuthPage renders the Register page correctly 1`] = `
<DocumentFragment>
  <div
    class="auth-page"
  >
    <div
      class="container page"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6 offset-md-3 col-xs-12"
        >
          <h1
            class="text-xs-center"
          >
            Sign up
          </h1>
          <p
            class="text-xs-center"
          >
            <a
              href="/login"
            >
              Have an account?
            </a>
          </p>
          <form>
            <fieldset
              class="form-group"
            >
              <input
                aria-label="Username"
                class="form-control form-control-lg"
                placeholder="Username"
                required=""
                type="text"
              />
            </fieldset>
            <fieldset
              class="form-group"
            >
              <input
                aria-label="Email"
                class="form-control form-control-lg"
                placeholder="Email"
                required=""
                type="email"
              />
            </fieldset>
            <fieldset
              class="form-group"
            >
              <input
                aria-label="Password"
                class="form-control form-control-lg"
                minlength="8"
                pattern=".{8,}"
                placeholder="Password"
                required=""
                type="password"
              />
            </fieldset>
            <button
              class="btn btn-lg btn-primary pull-xs-right"
              disabled=""
            >
              Sign up
              <svg
                id="loading-spinner"
                style="margin-left: 0.5rem; display: none;"
                viewBox="0 0 100 100"
                width="1rem"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="50"
                  cy="50"
                  id="circle"
                  r="45"
                  style="stroke: #fff;"
                />
              </svg>
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
